### DISCUSSION TITLE
Weird conflicting when try to impl `ToSql` in generic backend, with temporary value.

### DISCUSSION THREAD

**Discussion by visig9** [diesel-rs/diesel#4097] [Q&A] [answered]
I expect following code should work because `Sqlite` backend doesn't match the trait bound (`DB: Backend<BindCollector<'a> = RawBytesBindCollector<DB>>`). But it still conflicting. Don't known why.

# Source Code

`lib.rs`:

```rust
use diesel::{
    backend::Backend,
    expression::AsExpression,
    query_builder::bind_collector::RawBytesBindCollector,
    serialize::{self, IsNull, Output, ToSql},
    sql_types::Text,
    sqlite::Sqlite,
};
use std::fmt::{self, Display};

#[derive(AsExpression)]
#[diesel(sql_type = Text)]
struct Foo {
    id: String,
}

impl Display for Foo {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "I am {}", self.id)
    }
}

impl<DB> ToSql<Text, DB> for Foo
where
    for<'a> DB: Backend<BindCollector<'a> = RawBytesBindCollector<DB>>,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> serialize::Result {
        let s = self.to_string();
        <String as ToSql<Text, DB>>::to_sql(&s, &mut out.reborrow())
    }
}

impl ToSql<Text, Sqlite> for Foo {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
        out.set_value(self.to_string());
        Ok(IsNull::No)
    }
}
```

Error message:

```
$ cargo check
    Checking diesel-conflict v0.1.0 (/Users/ikai.wang/test/diesel-conflict)
error[E0119]: conflicting implementations of trait `ToSql<diesel::sql_types::Text, Sqlite>` for type `Foo`
  --> src/lib.rs:33:1
   |
23 | / impl<DB> ToSql<Text, DB> for Foo
24 | | where
25 | |     for<'a> DB: Backend<BindCollector<'a> = RawBytesBindCollector<DB>>,
   | |_______________________________________________________________________- first implementation here
...
33 |   impl ToSql<Text, Sqlite> for Foo {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo`

For more information about this error, try `rustc --explain E0119`.
error: could not compile `diesel-conflict` (lib) due to 1 previous error
```

`Cargo.toml`:

```toml
[package]
name = "diesel-conflict"
version = "0.1.0"
edition = "2021"

[dependencies]
diesel = { version = "2.2.1", features = ["sqlite"] }
```

# More Details

Basically I just copy the function signature from diesel source code here:
https://docs.rs/diesel/2.2.1/src/diesel/type_impls/primitives.rs.html#142-151

```rust
impl<DB> ToSql<sql_types::Text, DB> for str
where
    for<'a> DB: Backend<BindCollector<'a> = RawBytesBindCollector<DB>>,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> serialize::Result {
        out.write_all(self.as_bytes())
            .map(|_| IsNull::No)
            .map_err(|e| Box::new(e) as Box<dyn Error + Send + Sync>)
    }
}
```

And here:
https://docs.rs/diesel/2.2.1/src/diesel/sqlite/types/mod.rs.html#106-112

```rust
#[cfg(feature = "sqlite")]
impl ToSql<sql_types::Text, Sqlite> for str {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
        out.set_value(self);
        Ok(IsNull::No)
    }
}
```

Any idea? Thanks help.

**Comment by weiznich** [comment]
I'm not sure what answer you expect here. We as diesel developers cannot fix such compiler issues, so if the compiler complains about that some impl is conflicting there is nothing we can do to solve that.

That written: I agree with your assessment. I would expect as well that these impls are not conflicting. I vaguely remember that there are issues in the compiler with these higher ranked trait bounds, but for more details you need to reach out to the compiler team. Or to write it the other way around: You likely should fill an issue at the rustc repo for that.

**Reply by visig9** [reply]
Hi weiznich, thanks all your words! It's really helpful. I never thought this is a compiler bug before and just guess some subtle bugs has hidden in my code sample. I already search a while but maybe not used correct keywords.

Add more context: I'm trying to build some crates contains some basic data structure (like `Foo` here) and want to let them optional support `diesel` crate. So here is some features in `Cargo.toml`:

```toml
[features]
diesel = ["dep:diesel"]
diesel-sqlite = ["diesel/sqlite"]
``` 

`diesel::sqlite::Sqlite` is a special case and look like I have no way to bypass them when `ToSql` use temporary value. But If it's possible, I don't want to force user deal with more features like `diesel-postgres` or `diesel-mysql`, just because this kind of code:

```rust
impl ToSql<Text, Pg> for Foo {  // use `Pg` here so `diesel/postgres` is required.
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> serialize::Result {
        todo!()
    }
}
```

I almost give up but suddenly found `diesel` rely on the same(?) code snippet and not cause conflicting, but I can't reproduce that behavior in my crate. Maybe `diesel` using some technique which can sidestep the compiler bug? If you know that can you share it? Or have other suggestion in this scenario? Currently I'm using "more features" approach and going forward.

Anyway your answer is pretty useful already. Thanks for help!

**Reply by weiznich** [reply, marked as answer]
Well it might be that rustc does not consider that to be conflicting impl in the crate that owns the traits. That usually relaxes the rules quite a bit.

I do have currently no good suggestion on how you can deal with that without having separate feature flags for each backend.

**Reply by visig9** [reply]
No problems. Thanks you so much.
